controllers/cruisecontroloperation_controller.go

```diff
@@ -0,0 +1,446 @@
+	case banzaiv1alpha1.OperationRemoveBroker:
+		cruseControlTaskResult, err = r.Scaler.RemoveBrokersWithParams(ccOperationExecution.GetCurrentTaskParameters())
+	case banzaiv1alpha1.OperationRebalance:
+		cruseControlTaskResult, err = r.Scaler.RebalanceWithParams(ccOperationExecution.GetCurrentTaskParameters())
+	case banzaiv1alpha1.OperationStopExecution:
+		cruseControlTaskResult, err = r.Scaler.StopExecution()
+	default:
+		err = errors.NewWithDetails("Cruise Control operation not supported", "name", ccOperationExecution.GetName(), "namespace", ccOperationExecution.GetNamespace(), "operation", ccOperationExecution.GetCurrentTaskOp(), "parameters", ccOperationExecution.GetCurrentTaskParameters())
+	}
+	return cruseControlTaskResult, err
+}
+
+func sortOperations(ccOperations []*banzaiv1alpha1.CruiseControlOperation) map[string][]*banzaiv1alpha1.CruiseControlOperation {
+	ccOperationQueueMap := make(map[string][]*banzaiv1alpha1.CruiseControlOperation)
+	for _, ccOperation := range ccOperations {
+		switch {
+		case isWaitingforFinalize(ccOperation):
+			ccOperationQueueMap[ccOperationForFinalize] = append(ccOperationQueueMap[ccOperationForFinalize], ccOperation)
+		case ccOperation.IsWaitingForFirstExecution():
+			ccOperationQueueMap[ccOperationFirstExecution] = append(ccOperationQueueMap[ccOperationFirstExecution], ccOperation)
+		case ccOperation.IsWaitingForRetryExecution():
+			ccOperationQueueMap[ccOperationRetryExecution] = append(ccOperationQueueMap[ccOperationRetryExecution], ccOperation)
+		case ccOperation.IsInProgress():
+			ccOperationQueueMap[ccOperationInProgress] = append(ccOperationQueueMap[ccOperationInProgress], ccOperation)
+		}
+	}
+
+	// Sorting by operation type and by the k8s object creation time
+	for key := range ccOperationQueueMap {
+		sort.SliceStable(ccOperationQueueMap[key], func(i, j int) bool {
+			ccOperationQueue := ccOperationQueueMap[key]
+			if executionPriorityMap[ccOperationQueue[i].GetCurrentTaskOp()] < executionPriorityMap[ccOperationQueue[j].GetCurrentTaskOp()] {
+				return true
+			} else if executionPriorityMap[ccOperationQueue[i].GetCurrentTaskOp()] == executionPriorityMap[ccOperationQueue[j].GetCurrentTaskOp()] {
+				return ccOperationQueue[i].CreationTimestamp.Unix() < ccOperationQueue[j].CreationTimestamp.Unix()
+			}
+			return true
+		})
+	}
+	return ccOperationQueueMap
+}
+
+func selectOperationForExecution(ccOperationQueueMap map[string][]*banzaiv1alpha1.CruiseControlOperation) *banzaiv1alpha1.CruiseControlOperation {
+	// SELECTING OPERATION FOR EXECUTION
+	var ccOperationExecution *banzaiv1alpha1.CruiseControlOperation
+	// First prio: execute the finalize task
+	switch {
+	case len(ccOperationQueueMap[ccOperationForFinalize]) > 0:
+		ccOperationExecution = ccOperationQueueMap[ccOperationForFinalize][0]
+		ccOperationExecution.GetCurrentTask().Operation = banzaiv1alpha1.OperationStopExecution
```

Comment:
Not quite sure but can it happen that a Finalize task is already running and we stop that to start another?

Reply:
Good question. I did not think about this case. It should be tested.

Reply:
I have tested. Stop proposal execution runs instantly and the interrupted task will get completedWithError state.
Not possible to interrupt the finalizing task.